iceoryx_posh/test/integrationtests/test_service_discovery.cpp

```diff
@@ -473,31 +494,130 @@ TEST_F(ServiceDiscovery_test, FindServiceReturnsMaxServices)
-TEST_F(ServiceDiscovery_test, FindServiceReturnsContainerOverflowErrorWhenMoreThanMaxServicesAreFound)
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsAttachableToWaitSet)
 {
-    ::testing::Test::RecordProperty("TEST_ID", "f2f8d8c0-8712-4e7a-9e33-2b2a918f8a71");
-    const IdString_t SERVICE = "s";
-    const size_t numberOfInstances = (iox::MAX_NUMBER_OF_SERVICES + 1);
-    iox::cxx::vector<iox::popo::UntypedPublisher, numberOfInstances> publishers;
-    for (size_t i = 0; i < numberOfInstances; i++)
+    ::testing::Test::RecordProperty("TEST_ID", "fc0eeb7a-6f2a-481f-ae8a-1e17460e261f");
+    iox::popo::WaitSet<10U> waitSet;
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(testCallback))
+        .and_then([]() { GTEST_SUCCEED(); })
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsNotifiedbyWaitSetAboutSingleService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f1cf36b5-3db2-4e6f-8e05-e7e449530ec0");
+    iox::popo::WaitSet<1U> waitSet;
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(testCallback))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION("Moep", "Fluepp", "Shoezzel");
+    iox::popo::UntypedPublisher publisher(SERVICE_DESCRIPTION);
+
+    auto notificationVector = waitSet.wait();
+
+    for (auto& notification : notificationVector)
     {
-        std::string instance = "i" + iox::cxx::convert::toString(i);
-        iox::capro::ServiceDescription SERVICE_DESCRIPTION(
-            SERVICE, IdString_t(iox::cxx::TruncateToCapacity, instance), "foo");
-        publishers.emplace_back(SERVICE_DESCRIPTION);
+        (*notification)();
     }
 
-    auto serviceContainer = sut.findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard);
+    EXPECT_TRUE(callbackWasCalled);
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryNotifiedbyWaitSetFindsSingleService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1ecde7e0-f5b2-4721-b309-66f32f40a7bf");
+    iox::popo::WaitSet<1U> waitSet;
+    iox::capro::ServiceDescription serviceDescriptionToSearchFor("Soep", "Moemi", "Luela");
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(searchForService, serviceDescriptionToSearchFor))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+
+    iox::popo::UntypedPublisher publisher(serviceDescriptionToSearchFor);
+
+    auto notificationVector = waitSet.wait();
```

Comment:
In theory this can deadlock here and block the tests forever. Strictly speaking we would require a watchdog here a well. In particular the other thread from the watchdog can unblock the wait with `waitset.markForDestruction();`(strange name but still). In this case the test would fail but subsequent tests could continue.

Reply:
True, I added a timeout parameter to the call of `timedWait`.